helpers/checkpoint_logger/__init__.py

```diff
@@ -426,21 +428,29 @@ def submit_subflow(self: _Self, metric: str, start: T, end: T) -> _Self:
             sentry_sdk.set_measurement(metric, duration, "milliseconds")
             duration_in_seconds = duration / 1000
             PROMETHEUS_HANDLER.log_subflow(
-                flow=self.cls.__name__, subflow=metric, duration=duration_in_seconds
+                flow=self.cls.__name__,
+                subflow=metric,
+                duration=duration_in_seconds,
+                context=self.data.get("context"),
             )
 
         return self
 
 
 def from_kwargs(
-    cls: type[T], kwargs: MutableMapping[str, Any], strict: bool = False
+    cls: type[T],
+    kwargs: MutableMapping[str, Any],
+    strict: bool = False,
 ) -> CheckpointLogger[T]:
+    context = kwargs.pop("context", CheckpointContext())
     data = kwargs.get(_kwargs_key(cls), {})
 
     # kwargs has been deserialized into a Python dictionary, but our enum values
     # are deserialized as simple strings. We need to ensure the strings are all
     # proper enum values as best we can, and then downcast to enum instances.
     deserialized_data = {}
+    deserialized_data["context"] = context
+
```

Comment:
there's a subtle bug here. you have to do `data.pop("context")` because otherwise the for loop later in this function that iterates over `data.items()` will explode when trying to cast `"context"` to an `UploadFlow` enum (or whichever flow). also i think `kwargs.pop("context")` will mutate the passed-in kwargs which might not be safe

taking a step back, i think we don't have a tidy way to manage this context. if we store it in `self.data`, then each flow's `self.data` has its own copy and that's redundant. if we put a single copy in various tasks' `kwargs`, we have to remember to manually pass it and make sure it's included in kwargs when we retry and yadda yadda. since we want to do similar things for `log.info()` calls (task_name, task_id) and SQL metrics (repo_id, owner_id, commit_sha), i wrote https://github.com/codecov/worker/pull/810 which i think can neatly take care of it for all of them